kubeflow/trainer/backends/training_runtime_loader.py

```diff
@@ -0,0 +1,657 @@
+# Copyright 2025 The Kubeflow Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Shared runtime loader for container backends (Docker, Podman).
+
+We support loading training runtime definitions from multiple sources:
+1. GitHub: Fetches latest runtimes from kubeflow/trainer repository (with caching)
+2. Local bundled: Falls back to `kubeflow/trainer/config/training_runtimes/` YAML files
+3. User custom: Additional YAML files in the local directory
+
+The loader tries GitHub first (with 24-hour cache), then falls back to bundled files
+if the network is unavailable or GitHub fetch fails.
+"""
+
+from __future__ import annotations
+
+from datetime import datetime, timedelta
+import json
+import logging
+from pathlib import Path
+from typing import Any
+import urllib.error
+import urllib.request
+
+import yaml
+
+from kubeflow.trainer.constants import constants
+from kubeflow.trainer.types import types as base_types
+
+logger = logging.getLogger(__name__)
+
+TRAINING_RUNTIMES_DIR = Path(__file__).parents[1] / "config" / "training_runtimes"
+CACHE_DIR = Path.home() / ".kubeflow" / "trainer" / "cache"
+CACHE_DURATION = timedelta(hours=24)
+
+# GitHub runtimes configuration
+GITHUB_RUNTIMES_BASE_URL = (
+    "https://raw.githubusercontent.com/kubeflow/trainer/master/manifests/base/runtimes"
+)
+GITHUB_RUNTIMES_TREE_URL = "https://github.com/kubeflow/trainer/tree/master/manifests/base/runtimes"
+
+
+def _load_runtime_from_yaml(path: Path) -> dict[str, Any]:
```

Comment:
Why do you want to keep these functions private if all functions in this file are private ?
```suggestion
def load_runtime_from_yaml(path: Path) -> dict[str, Any]:
```